pkg/validation/attributes/attributes.go

```diff
@@ -0,0 +1,62 @@
+package attributes
+
+import (
+	"regexp"
+	"strings"
+
+	"github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
+	apiAttributes "github.com/devfile/api/v2/pkg/attributes"
+)
+
+// ValidateGlobalAttribute validates the workspace template spec data for global attribute references
+func ValidateGlobalAttribute(workspaceTemplateSpec *v1alpha2.DevWorkspaceTemplateSpec) error {
+
+	var err error
+
+	if workspaceTemplateSpec != nil {
+		// Validate the components
+		if err = ValidateComponents(workspaceTemplateSpec.Attributes, &workspaceTemplateSpec.Components); err != nil {
+			return err
+		}
+
+		// Validate the commands
+		if err = ValidateCommands(workspaceTemplateSpec.Attributes, &workspaceTemplateSpec.Commands); err != nil {
+			return err
+		}
+
+		// Validate the events
+		if err = ValidateEvents(workspaceTemplateSpec.Attributes, workspaceTemplateSpec.Events); err != nil {
+			return err
+		}
+
+		// Validate the projects
+		if err = ValidateProjects(workspaceTemplateSpec.Attributes, &workspaceTemplateSpec.Projects); err != nil {
+			return err
+		}
+
+		// Validate the starter projects
+		if err = ValidateStarterProjects(workspaceTemplateSpec.Attributes, &workspaceTemplateSpec.StarterProjects); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+var globalAttributeRegex = regexp.MustCompile(`\{{2}(.*?)\}{2}`)
+
+// validateAndReplaceDataWithAttribute validates the string for a global attribute and replaces it. An error
+// is returned if the string references an invalid global attribute key
+func validateAndReplaceDataWithAttribute(val string, attributes apiAttributes.Attributes) (string, error) {
+	matches := globalAttributeRegex.FindAllStringSubmatch(val, -1)
+	for _, match := range matches {
+		var err error
+		attrValue := attributes.GetString(match[1], &err)
+		if err != nil {
+			return "", err
+		}
```

Comment:
Are we sure we want to consider this an error case? It basically forbids using `{{text}}` anywhere in a Devfile. I think a safer choice is to leave it unmodified.

Reply:
Hmm interesting thought. So if we go by the suggested convention, it either replaces the valid `{{key}}` to its value or leaves the invalid key `{{invalidKey}}` as it is? Then there would be no error cases at all.. 

What happens when someone commits an unintentional typo or it maybe confusing if an error is not returned since we are allowing the global attributes to overriden and merged 🤔

Reply:
Yeah, not requiring replacement does leave the ugliness of trying to execute something like `go{{go-version}}` (though ideally one would notice that pretty quickly). My main worry is that there might be a use case _somewhere_ that requires `{{text}}` and this approach would effectively make us incompatible (e.g. I know helm charts are littered with `{{}}`). 

Good point about parent attributes though -- I think the _most_ confusing outcome would be an attribute from parent replacing `{{my-var}}` in your code despite you never seeing the `my-var` attribute defined. I'm pretty split here.

Reply:
@amisevsk regd the parent case, I think if a user is intending to use a parent devfile, they should be aware of the possible consequences of parent data being merged in the main devfile content, in this case a parent global attribute `my-var` replacing all occurances of `{{my-var}}` in the final devfile.

Reply:
@elsony what are your thoughts about future incompatibility arising from using `{{}}` in devfile for global attributes because like @amisevsk says Helm charts uses this convention.

Reply:
@sleshchenko not sure I follow, are we saying convention should be one pair `{}` ie,  `{var}`?

@amisevsk I thought about a bit more, I am not sure how devfile data will give way to incompatibility with a Helm chart. `{{my-var}}` in a devfile is either resolved or not. If its not resolved, we can err out(also depends upon our other conversation whether we should err out) or it will remain as `{{my-var}}`. If this data is being used in Helm, it will most likely not follow the Helm convention which is something like `{{.IMAGE}}`(notice the period) so in those cases will likely result in a Helm failure 🤔

Reply:
I've thought about it a bit more and I'm going towards "this shouldn't be an error". Failing here causes subtle complexity that's hard to resolve, e.g. the devfiles:
```yaml
metadata:
  name: parent-devfile
attributes:
  parent-attribute: "test"
components:
  - name: parent-component
    container:
      image: quay.io/test/image
---
metadata:
  name: main-devfile
attributes:
  parent-devfile-id: "parent-devfile"
parent:
  id: "{{parent-devfile-id}}"
components:
  - name: child-component
    image: quay.io/{{parent-attribute}}/image
```
become impossible to reconcile: doing the replacement in `main-devfile` will fail because `{{parent-attribute}}` does not exist, and we can't flatten in the parent attribute becuase its id is hidden behind `{{parent-devfile-id}}`, which requires attribute replacement.

Reply:
@amisevsk I've explained the steps in this comment here regd this case https://github.com/devfile/api/pull/352#discussion_r587591297 pls do share your thoughts